### PR TITLE
Refactor cache and correct type hints

### PR DESCRIPTION
**Pull request by tisnik**
## Type of change

- [x] Refactor
- [ ] New feature
- [x] Bug fix
- [ ] Optimization
- [ ] Documentation Update

## Description

- `Cache` class in separate source file to avoid cycles during import
- Fixed type hints - now `RedisCache` and `InMemoryCache` are really inherited from `Cache`
- That fixed `CacheFactory.conversation_cache` type issue
- Two new unit tests for `CacheFactory` to check the behaviour

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.

## Testing
- Partially tested by unit tests (`InMemoryCache`)
- Added two new unit tests for `CacheFactory` too

### CODE REVIEW

File: src/cache/in_memory_cache.py [hunk 1]
```diff
@@ -2,9 +2,10 @@
 from typing import Union
 import threading
 import src.constants as constants
+from src.cache.cache import Cache
 
 
-class InMemoryCache:
+class InMemoryCache(Cache):
```
Comment (vishnuchalla) [inline src/cache/in_memory_cache.py:8]: Thanks for catching this.

**Comment by vishnuchalla** [comment]
changes wise, looks good to me. I will let others to review and merge it. Thank you

**Comment by bparees** [comment]
/approve

thanks!

**Pull request merged**